// Test.QuickCheck.Data.AlphaNumString/index.js
"use strict";
var Prelude = require("../Prelude");
var Data_Newtype = require("../Data.Newtype");
var Data_String = require("../Data.String");
var Test_QuickCheck_Gen = require("../Test.QuickCheck.Gen");
var Test_QuickCheck_Arbitrary = require("../Test.QuickCheck.Arbitrary");
var Data_Eq = require("../Data.Eq");
var Data_Ord = require("../Data.Ord");
var Control_Applicative = require("../Control.Applicative");
var Data_Functor = require("../Data.Functor");
var Control_Semigroupoid = require("../Control.Semigroupoid");

/**
 *  | A newtype for `String` whose `Arbitrary` instance generated random
 *  | alphanumeric strings.
 */
var AlphaNumString = function (x) {
    return x;
};
var ordAlphaNumString = Data_Ord.ordString;
var newtypeAlphaNumString = new Data_Newtype.Newtype(function (n) {
    return n;
}, AlphaNumString);
var eqAlphaNumString = Data_Eq.eqString;
var coarbAlphaNumString = Test_QuickCheck_Arbitrary.coarbString;
var arbAlphaNumString = new Test_QuickCheck_Arbitrary.Arbitrary((function () {
    var rest = Data_String.toCharArray("bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
    var anyChar = Test_QuickCheck_Gen.oneOf(Control_Applicative.pure(Test_QuickCheck_Gen.applicativeGen)("a"))(Data_Functor.map(Data_Functor.functorArray)(Control_Applicative.pure(Test_QuickCheck_Gen.applicativeGen))(rest));
    return Data_Functor.map(Test_QuickCheck_Gen.functorGen)(function ($3) {
        return AlphaNumString(Data_String.fromCharArray($3));
    })(Test_QuickCheck_Gen.arrayOf(anyChar));
})());
module.exports = {
    AlphaNumString: AlphaNumString, 
    newtypeAlphaNumString: newtypeAlphaNumString, 
    eqAlphaNumString: eqAlphaNumString, 
    ordAlphaNumString: ordAlphaNumString, 
    arbAlphaNumString: arbAlphaNumString, 
    coarbAlphaNumString: coarbAlphaNumString
};
